### DISCUSSION TITLE
Centralized org-wide linter configs? I'm so close to getting it working!

### DISCUSSION THREAD

**Discussion by BretFisher** [super-linter/super-linter#2640] [Ideas]
Long-time Super-Linter user here, hoping for some ideas on working around a current limitation of the project when going for a reusable workflow **with** centralized linter configs (configs in a single repo for use in many repos):

Now with reusable workflows getting an increased feature set like `internal` repo support, calling workflows in the same repo, and other advancements, I'm working across multiple projects trying to implement Super-Linter in larger (10-30 repos) teams.  One real challenge is we end up with 5-10 `.github/linters` override files that still have to be in every repo, even if we can use super-linter from a central reusable workflow.

I've tried to implement an extra step in my super-linter workflow that clones the reusable workflow repo and copies in linter configs, but I've discovered that sometimes super-linter changes branches and causes an error due to uncommitted files:

I've added these steps right before super-linter runs. They checkout the central super-linter repo as a sub-dir, copy in linter configs, then delete that repo:

```yaml
      - name: Clone org/super-linter for default linter configs
        uses: actions/checkout@v3.0.0
        with:
          repository: 'org/super-linter'
          ref: 'main'
          path: 'super-linter'
          clean: 'false'
          persist-credentials: 'false'
          ssh-key: '${{ secrets.gha-deploy-key }}'

      - name: Copy linter config defaults from org/super-linter
        run: |
          cp --recursive --no-clobber super-linter/.github/linters/ .github/linters/
          ls -al .github/linters/
          rm -rf super-linter
```

This worked fantastic at first in simple repos, but once I started adding it to more complex existing repos, I get this error in a super-linter step. I understand that git can't be dirty when switching branches, but **was hoping someone had an idea for working around these limitations, or maybe a super-linter feature that provides a way to implement central configs**:

```text
2022-03-15 01:08:30 [INFO]   Gathering user validation information...
2022-03-15 01:08:30 [INFO]   - Only validating [new], or [edited] files in code base...
2022-03-15 01:08:32 [INFO]   ----------------------------------------------
2022-03-15 01:08:32 [INFO]   User provided file:[/github/workspace/.github/linters/.hadolint.yaml] exists, setting rules file...
2022-03-15 01:08:32 [INFO]   ----------------------------------------------
2022-03-15 01:08:32 [INFO]   User provided file:[/github/workspace/.github/linters/actionlint.yml] exists, setting rules file...
2022-03-15 01:08:32 [INFO]   ----------------------------------------------
2022-03-15 01:08:32 [INFO]   User provided file:[/github/workspace/.github/linters/.gitleaks.toml] exists, setting rules file...
2022-03-15 01:08:32 [INFO]   ----------------------------------------------
2022-03-15 01:08:32 [INFO]   User provided file:[/github/workspace/.github/linters/.markdown-lint.yml] exists, setting rules file...
2022-03-15 01:08:33 [INFO]   ----------------------------------------------
2022-03-15 01:08:33 [INFO]   User provided file:[/github/workspace/.github/linters/.tflint.hcl] exists, setting rules file...
2022-03-15 01:08:34 [INFO]   ----------------------------------------------
2022-03-15 01:08:34 [INFO]   User provided file:[/github/workspace/.github/linters/terrascan.toml] exists, setting rules file...
2022-03-15 01:08:34 [INFO]   ----------------------------------------------
2022-03-15 01:08:34 [INFO]   User provided file:[/github/workspace/.github/linters/.yaml-lint.yml] exists, setting rules file...
You are not currently on a branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

2022-03-15 01:08:35 [INFO]   Failed to switch to main branch to get files changed!
2022-03-15 01:08:35 [FATAL]   [error: The following untracked working tree files would be overwritten by checkout:
	.github/linters/.gitleaks.toml
	.github/linters/.hadolint.yaml
	.github/linters/.tflint.hcl
	.github/linters/actionlint.yml
	.github/linters/terrascan.toml
Please move or remove them before you switch branches.
Aborting]
```

**Comment by BretFisher** [comment]
Oh, I'm just now seeing `LINTER_RULES_PATH`, and thought that was always hardcoded. I'll try copying linters out of the git path.

**Comment by BretFisher** [comment]
OK spent some time on this and `LINTER_RULES_PATH` isn't solving the core issue, which is Super-Linter having no overrides for the project root when using GHAs.

In a GitHub Runner, Super-Linter seems to only allow the repo to be in the root work directory. If I checkout the main code to any subdir, Super-Linter errors saying it's not a git project.  The `DEFAULT_WORKSPACE` option doesn't seem to work in GHA.

If I clone/copy my linter defaults into that path, then git is dirty and any git branch change on Super-Linter's behalf will error.

I've tried doing the main `actions/checkout` to a subdir, then doing the central linter configs `actions/checkout` to a different subdir. Then combining their linter configs in a 3rd subdir. All three subdirs are in the default Runner workdir, and this avoids changing anything in the main repo's checkout, and I can use `LINTER_RULES_PATH` to point to that 3rd subdir for configs, but **Super-Linter will still fail because there is no .git in workdir and I can't seem to change the workdir for Super-Linter.**

I then thought about augmenting the global .gitignore on each run to temporarily ignore a new linter config directory, but it feels like I'm down a deep rabbit hole just to be able to use remote linter configs in Super-Linter.  😄

**Reply by BretFisher** [reply]
@admiralAwkbar any ideas on how I might achieve this, or a super-linter feature that might help this work? Like the ability to checkout code in a sub-dir that we can focus super-linter on as the WORKSPACE?

**Comment by mohsyed04** [comment]
@BretFisher Can you please share how you got it working for the whole repository, I'm only able to run it for branches listed under 
on:
    pull_request:
          [...., branch names, .....]

**Reply by BretFisher** [reply]
"Whole repository" I'm not sure what you mean. Can you explain what you're trying to do in more detail?

**Reply by mohsyed04** [reply]
I have this file:

 ```
pull_request:
    branches: [next]

jobs:
  build:
    name: Lint Code Base
    runs-on: ubuntu-latest
steps:
      - name: Checkout Code
        uses: actions/checkout@v3
        with:
          fetch-depth: 0

      - name: Lint Code Base
        uses: github/super-linter@v4
        env:
          VALIDATE_ALL_CODEBASE: false
          DEFAULT_BRANCH: next
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          DISABLE_ERRORS: true
          
          PYTHON_FLAKE8_CONFIG_FILE: .flake8
          VALIDATE_PYTHON_FLAKE8: true`
```
 It only runs on the branches listed under "branches" parameter, so either I have to list all the branches in the repository that I want it to run on, or edit the file on each branch with respective branch name. I want it to work on all branches in the repository without having to edit that file on every branch.
I was wondering if I can match/list all branches by using some pattern matching (like **,  all., etc)

**Comment by BretFisher** [comment]
Remove the branches line. Then it runs in all PRs regardless of branch.

**Reply by mohsyed04** [reply]
Thanks, got it working👍
